### ISSUE TITLE
Improve search UX by adding a loader while initializing the search

### ISSUE THREAD

**Issue by tmkasun** [squidfunk/mkdocs-material#1439]
__I checked that...__

- [x] ... the documentation does not mention anything about my idea
- [x] ... to my best knowledge, my idea wouldn't break something for other users
- [x] ... there are no open or closed issues that are related to my idea

## Description

Add a loading animation to search input field until the search_index file gets loaded and parsed

### Use Cases

When the `search_index.json` file is large for larger document sites, It takes a while to load and parse the file in the browser, Until it completes this process, User feels like search operation is not working, because we don't give any feedback until the search initialization complete (On click search input, loads `search_index.json`, parse it)

### Screenshots / Mockups

[This is how](https://github.com/tmkasun/mkdocs-material/commit/3e92e6cd0c72bbee571cacd6941c87d25e3dd6aa#diff-8d5b767cb91a4401d3a9735774b182b3R164-R177) we have customized search initialization in our docs

![Jan-26-2020 22-01-02](https://user-images.githubusercontent.com/3313885/73138332-6b1c4980-4087-11ea-9298-a14341ffe652.gif)

**Comment by squidfunk** [comment]
This is indeed a good idea from a UX perspective but probably not necessary anymore with the refactoring I‘ve been doing in #1306. Search now runs in a web worker and the index is built upon page load and not on the first focus. Furthermore, prebuilt indexes are supported and the index is also persisted in local storage. This should make search instantaneous. 

Let‘s see how the new implementation performs in the wild (when v5 is out). If this is still a problem, we can add a loading indicator.

**Comment by tmkasun** [comment]
@squidfunk , Obviously, worker approach will decrease the time to become interactive and will not block the main thread. But IMHO even though it parallelizes the fetch & initialization operations, there will be a delay in IO and processing inside the worker. Hence I think, Search functionality will not be ready immediately until the worker finishes the initialization.
anyway I will try  v5 preview 2 and check the how it works with our scenarios

**Comment by squidfunk** [comment]
@tmkasun there are more changes in the `refactor/typescript-rxjs` branch, so it's probably best to try that. I'll issue a new preview release the next days and a RC when I finished migrating everything to the new architecture.

**Comment by tmkasun** [comment]
@squidfunk , Thanks , Will try `refactor/rxjs-typescript` branch

**Comment by squidfunk** [comment]
Just released 5.0.0b1 which includes the new search functionality, see #1465 - any feedback is appreciated

**Comment by squidfunk** [comment]
This was closed accidentally.

**Comment by wilhelmer** [comment]
Related bug report from https://github.com/squidfunk/mkdocs-material/issues/1498#issuecomment-609585090:

> In v4, if you started searching before `search_index.json` was loaded, search results were displayed as soon as the index had been loaded. 
> 
> In v5, if you start searching before the index is loaded, the search displays "No matching results" and keeps displaying that message even after the index has been loaded. 
> 
> For example, if you type "test" and wait until the search index is loaded, Material keeps displaying "No matching results" until you press a key. You must change the search term to, e.g., "tes" and then back to "test" to get your search results.
> 
> Steps to reproduce:
> 
> 1. Visit https://docs.baslerweb.com
> 2. Enable DevTools
> 3. Disable cache and throttle to Fast 3G or slower
> 4. Page refresh
> 5. Start searching for "test" immediately

If you implement a loading indicator, it would be great if you could also make search results appear as soon as loading is done.

**Comment by wilhelmer** [comment]
@squidfunk Is there any in-built way to be notified when the search index has been fully loaded? I know there's `app.search$.subscribe(({ query }) => console.log(query))`, but that works only after the user has entered a query. I need something that works right on page load without any user interaction.

**Comment by squidfunk** [comment]
Currently not, but we could add it when adding the loading indicator.

**Comment by wilhelmer** [comment]
Maybe add that first so we can easily build our own loading indicator.

**Comment by squidfunk** [comment]
Search needs a small refactoring as some people wish to load the search index on the first focus. We should add an option when search is initialized and a state member to the emitted values of the search observable denoting the state.

**Comment by wilhelmer** [comment]
Super simple loading indicator, for those who need a temporary quick-and-dirty solution:

```js
$(".md-search__input").on("input", function () {
    if (!app.search$.isLoaded) {
        $(".md-search-result__meta").text("Loading search index ...");
    }
});

app.search$.subscribe(function () {
    app.search$.isLoaded = true;
});
```

**Comment by squidfunk** [comment]
Nice one! Note that it relies on jQuery 😉

**Comment by wilhelmer** [comment]
Yes, heresy! Too lazy for vanilla on Sunday 😄

**Comment by squidfunk** [comment]
b5be7d39 adds a `status` member to the search observable. Search initialization can now be observed with:

``` js
app.search$.subscribe(({ status }) => {
  if (status === "ready") {
    console.log("Search is ready")
  }
})
```

Note that the status value is included in every emission. To perform an action once when the search is ready, you can just use a variable from the parent scope:

``` js
var ready = false
app.search$.subscribe(({ status }) => {
  if (status === "ready" && !ready) {
    console.log("Search is ready")
    ready = true
  }
})
```

**Comment by squidfunk** [comment]
With 62fcc9b8, the latest query will now automatically be re-emitted when search changes its status from `waiting` to `ready`. The only thing that's left now is the loading indicator, which I'm still unsure about how to implement. If search is loaded on page load, it is most likely ready, when the user enters a term. For this reason, an indicator as proposed by the OP is too distracting, as only 5% of people seem to use the search. When we add the possibility back to initialize search on focus, a loading indicator might be helpful. IMHO the best approach is to replace the text which shows "Type to start searching" with "Loading search index" or something similar.

**Comment by squidfunk** [comment]
0145f6d2 adds a hint that the search is being initialized. Again, the last value is re-emitted, when search is successfully initialized. We could add some more animations to it, but I think it's a start.

Tested with Fast 3G on the official docs:

![Ohne Titel](https://user-images.githubusercontent.com/932156/80313631-5d907000-87ec-11ea-9138-17a45819140c.gif)

**Comment by squidfunk** [comment]
Released as part of 5.1.2

**Comment by wilhelmer** [comment]
Small error in the German language file:

https://github.com/squidfunk/mkdocs-material/blob/5ecb30b4a3cc167c3ac645eaed36e2c18b2f669c/src/partials/language/de.html#L35

should be `search.result.initializer`

**Comment by squidfunk** [comment]
Thanks! Fixed in 671ec724.
